src/server/enterprise/server/api_server.go

```diff
@@ -59,69 +64,153 @@ func NewEnterpriseServer(env *serviceenv.ServiceEnv, etcdPrefix string) (ec.APIS
+		return nil, err
+	}
+	return &ec.HeartbeatResponse{}, nil
+}
+
 // Activate implements the Activate RPC
 func (a *apiServer) Activate(ctx context.Context, req *ec.ActivateRequest) (resp *ec.ActivateResponse, retErr error) {
 	a.LogReq(req)
 	defer func(start time.Time) { a.pachLogger.Log(req, resp, retErr, time.Since(start)) }(time.Now())
 
-	// Validate the activation code
-	expiration, err := license.Validate(req.ActivationCode)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error validating activation code")
-	}
-	// Allow request to override expiration in the activation code, for testing
-	if req.Expires != nil {
-		customExpiration, err := types.TimestampFromProto(req.Expires)
-		if err == nil && expiration.After(customExpiration) {
-			expiration = customExpiration
-		}
-	}
-	expirationProto, err := types.TimestampProto(expiration)
+	// Try to heartbeat before persisting the configuration
+	heartbeatResp, err := a.heartbeatToServer(ctx, req.LicenseServer, req.Id, req.Secret)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not convert expiration time \"%s\" to proto", expiration.String())
+		return nil, err
 	}
+
+	// If the test heartbeat succeeded, write the state and config to etcd
 	if _, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-		e := a.enterpriseToken.ReadWrite(stm)
-		// blind write
-		return e.Put(enterpriseTokenKey, &ec.EnterpriseRecord{
-			ActivationCode: req.ActivationCode,
-			Expires:        expirationProto,
-		})
+		if err := a.configCol.ReadWrite(stm).Put(configKey, &ec.EnterpriseConfig{
+			LicenseServer: req.LicenseServer,
+			Id:            req.Id,
+			Secret:        req.Secret,
+		}); err != nil {
+			return err
+		}
+
+		return a.enterpriseTokenCol.ReadWrite(stm).Put(enterpriseTokenKey, &ec.EnterpriseRecord{License: heartbeatResp.License})
 	}); err != nil {
 		return nil, err
 	}
 
-	// Wait until watcher observes the write
+	// Wait until watcher observes the write to the state key
 	if err := backoff.Retry(func() error {
 		record, ok := a.enterpriseTokenCache.Load().(*ec.EnterpriseRecord)
 		if !ok {
 			return errors.Errorf("could not retrieve enterprise expiration time")
 		}
-		expiration, err := types.TimestampFromProto(record.Expires)
-		if err != nil {
-			return errors.Wrapf(err, "could not parse expiration timestamp")
-		}
-		if expiration.IsZero() {
+		if record.License == nil {
 			return errors.Errorf("enterprise not activated")
 		}
 		return nil
-	}, backoff.RetryEvery(time.Second)); err != nil {
+	}, backoff.RetryEvery(100*time.Millisecond)); err != nil {
 		return nil, err
 	}
-	time.Sleep(time.Second) // give other pachd nodes time to observe the write
 
-	return &ec.ActivateResponse{
-		Info: &ec.TokenInfo{
-			Expires: expirationProto,
-		},
-	}, nil
+	return &ec.ActivateResponse{}, nil
 }
 
 // GetState returns the current state of the cluster's Pachyderm Enterprise key (ACTIVE, EXPIRED, or NONE), without the activation code
```

Comment:
Yeah this comment could be clearer - we return a fragment of the activation code, but we cut the signature out so it can't be replayed. This was an issue in 1.x releases because sales uses the enterprise code as a unique identifier for user metrics, but the whole code was exposed on a public API with no auth. So before this change anyone could drive-by a Hub cluster and get a valid enterprise license for years.